src/textual/message_pump.py

```diff
@@ -578,7 +578,16 @@ def _get_dispatch_methods(
             # Try decorated handlers first
             decorated_handlers = cls.__dict__.get("_decorated_handlers")
             if decorated_handlers is not None:
-                handlers = decorated_handlers.get(type(message), [])
+                handlers: list[tuple[Callable, dict[str, tuple[SelectorSet, ...]]]] = []
+                add_handlers = handlers.extend
+                # Allow for firing of handlers bound to a message higher up
+                # the inheritance tree.
+                for check_message in message.__class__.__mro__:
+                    # If we've hit something that isn't derived from Message
+                    # we can give up.
+                    if not issubclass(check_message, Message):
+                        break
+                    add_handlers(decorated_handlers.get(check_message, []))
```

Comment:
Could this lead to a handler being called more than once for a message?

i.e if you have an inheritance chain Foo -> Bar -> Baz (where Foo is the base class).  It looks like if the messages was a `Baz` it would call the handler 3 times?

Could very well have misinterpreted the code.

Reply:
I *might* be misinterpreting the question, but isn't the situation mentioned here what this test is about?

```python
class MessageSender(Widget):
    class Parent(Message):
        pass

    class Child(Parent):
        pass

    def post_parent(self) -> None:
        self.post_message(self.Parent())

    def post_child(self) -> None:
        self.post_message(self.Child())


async def test_fire_on_inherited_message() -> None:
    """Handlers should fire when descendant messages are posted."""

    posted: list[str] = []

    class InheritTestApp(App[None]):
        def compose(self) -> ComposeResult:
            yield MessageSender()

        @on(MessageSender.Parent)
        def catch_parent(self) -> None:
            posted.append("parent")

        @on(MessageSender.Child)
        def catch_child(self) -> None:
            posted.append("child")

        def on_mount(self) -> None:
            self.query_one(MessageSender).post_parent()
            self.query_one(MessageSender).post_child()

    async with InheritTestApp().run_test():
        pass

    assert posted == ["parent", "child", "parent"]
```

If the message is `MessageSender.Child` then `InheritTestApp.catch_child` and `InheritTestApp.catch_parent` get triggered; which is what I'd expect here and actually desire.

Reply:
Okay, after a wee chat this morning, the (valid) concern here is what happens in *this* situation (that is, someone decorates the *same* handler with messages that are parent/child related):

```python
async def test_fire_inherited_on_single_handler() -> None:
    """Test having parent/child messages on a single handler."""

    posted: list[str] = []

    class InheritTestApp(App[None]):
        def compose(self) -> ComposeResult:
            yield MessageSender()

        @on(MessageSender.Parent)
        @on(MessageSender.Child)
        def catch_either(self, event: MessageSender.Parent) -> None:
            posted.append(f"either {event.__class__.__name__}")

        def on_mount(self) -> None:
            self.query_one(MessageSender).post_parent()
            self.query_one(MessageSender).post_child()

    async with InheritTestApp().run_test():
        pass

    # TODO: Likely not this!
    assert posted == ["either Parent", "either Child", "either Child"]
```